### PR TITLE
Be more selective in test runs

### PR DESCRIPTION
**Pull request by thomwiggers**
This should speed up the builds significantly.

* Sets `PQCLEAN_ONLY_DIFF` on test runs.
* Reduces the number of iterations functest runs
* Makes sure the CI services set up their branch correctly
* Sets up CircleCI result reporting
* Fixes a `PyYAML` warning about `yaml.load`
* Changes the `permit_test` helper into a function decorator:
   * this allows to easily attach it to functions
   * This also makes sure the tests still show up in the build output, marked `skipped`

### CODE REVIEW

**Comment by thomwiggers** [comment]
Currently testing what this will do in my fork of PQClean.

**Comment by thomwiggers** [comment]
@dstebila If I understood the helper code correctly, on the master branch we should always build everything right?

https://dev.azure.com/ret0171/PQClean%20builds/_build/results?buildId=26 only builds aes.

**Comment by dstebila** [comment]
> @dstebila If I understood the helper code correctly, on the master branch we should always build everything right?

That's what I initially thought, but am open to alternatives.

> https://dev.azure.com/ret0171/PQClean%20builds/_build/results?buildId=26 only builds aes.
> 
> EDIT: Though it seems that azure pipelines only checks out a specific commit and does not select the branch.

I'm not sure how to respond to this part.

**Comment by thomwiggers** [comment]
> > @dstebila If I understood the helper code correctly, on the master branch we should always build everything right?
> 
> That's what I initially thought, but am open to alternatives.
> 
> > https://dev.azure.com/ret0171/PQClean%20builds/_build/results?buildId=26 only builds aes.
> > EDIT: Though it seems that azure pipelines only checks out a specific commit and does not select the branch.
> 
> I'm not sure how to respond to this part.
I meant this as a mental note to myself and to answer my own question about why this is happening:

* build is from the master branch
* checked out is a detached head of a commit on the master branch so the code doesn't realise it's on master
* difference between HEAD and origin/master is empty.

**Comment by thomwiggers** [comment]
This has been tested on my fork to be working fine now.

File: test/test_testvectors.py [hunk 1]
```diff
@@ -10,28 +10,27 @@
 
 
 def test_testvectors():
+    @helpers.filtered_test
+    def check_testvectors(implementation):
```
Comment (dstebila) [inline test/test_testvectors.py]: Why did you move this inside?

File: .circleci/config.yml [hunk 2]
```diff
@@ -4,40 +4,46 @@ version: 2
...
-          docker run -e CI=true --rm -v `pwd`:`pwd` -w `pwd` "pqclean/ci-container:$ARCH" /bin/bash -c "
+          docker run -e CI=true -e PQCLEAN_ONLY_DIFF=1 --rm -v `pwd`:`pwd` -w `pwd` "pqclean/ci-container:$ARCH" /bin/bash -c "
           uname -a &&
           export CC=${CC} &&
           pip3 install -r requirements.txt &&
-          cd test && python3 -m nose --rednose --verbose"
+          mkdir test-results &&
+          cd test && python3 -m nose --rednose --verbose --with-xunit --xunit-file=../test-results/nosetests.xml"
+    - store_test_results:
+        path: test-results
```
Comment (dstebila) [inline .circleci/config.yml]: Can you explain to me what this is doing?

Comment (thomwiggers) [reply .circleci/config.yml]: It produces slightly nicer result output on the CircleCI pages.  See test summary on https://circleci.com/gh/PQClean/PQClean/3161

File: test/test_testvectors.py (hunk 1 above)
Comment (thomwiggers) [reply test/test_testvectors.py]: To have it not be detected by nosetests as a test, because test is part of the name of this function. It used to be called `check_vectors`, but as the decorator filter uses the name of the function to determine if it should be run.
